File: src/classes/ArcTransPathPlanner.m
classdef ArcTransPathPlanner < handle
% plan a continuous path according to the given points
% arc transits between 2 lines
% ref: https://blog.csdn.net/qq_26565435/article/details/98789361
% ref: Robotics: Modelling, Plannig and Control, chapter 4.3
    
    properties
        center
        radius
        theta
        rot
        dr
        dl
        pt
        line_vec
        distance
        dis_interval
    
        nump
        numarc
        p_initial
        p_goal
        trans_type
    end
    
    
    methods
        %% Class Constructor
        function obj = ArcTransPathPlanner(pos, radius)
            obj.p_initial = pos(:,1);
            obj.p_goal = pos(:,end);
            obj.nump = size(pos, 2);
            if radius>0
                obj.numarc = obj.nump-2;
                obj.radius = radius;
                obj.CalcArcInfo(pos);
                obj.trans_type = 'arc';
            else
                obj.numarc = obj.nump/2-1;
                obj.radius = 0.5*norm(pos(:,3)-pos(:,2));
                obj.CalcSemicircleInfo(pos);
                obj.trans_type = 'semicircle';
            end
            obj.dis_interval = obj.CalcDisInsterval();
        end
        
        %% Calculate Arc Transition
        function CalcArcInfo(obj, pos)
            for idx=1:obj.numarc
                [obj.center(:,idx), obj.theta(idx), pt1, pt2] = obj.CalcArcPoints(pos(:,idx), pos(:,idx+1), pos(:,idx+2));
                obj.pt(:,2*idx-1) = pt1;
                obj.pt(:,2*idx) = pt2;
                obj.dr(idx) = obj.radius*(pi-obj.theta(idx));
                obj.dl(idx) = norm(pt1-pos(:,idx));
                obj.rot{idx} = obj.CalcArcRot(obj.center(:,idx), pt1, pt2);
                if idx ~=1
                    obj.dl(idx) = obj.dl(idx)-norm(pos(:,idx)-obj.pt(:,2*(idx-1)));
                end
                if idx==obj.numarc
                    obj.dl(idx+1) = norm(pos(:,end)-pt2);
                end
                obj.line_vec(:,idx) = (pos(:,idx+1)-pos(:,idx))/norm(pos(:,idx+1)-pos(:,idx));
            end
            obj.line_vec(:,obj.numarc+1) = (pos(:,end)-pos(:,end-1))/norm(pos(:,end)-pos(:,end-1));
            obj.distance = sum(obj.dl)+sum(obj.dr);
        end
        
        function [center, theta, pt1, pt2] = CalcArcPoints(obj, p1, p2, p3)
            p21 = p1-p2; p23 = p3-p2;
            tmp_cos = dot(p21, p23)/norm(p21)/norm(p23);
            theta = acos(tmp_cos);
            
            p2t1 = obj.radius/tan(0.5*theta)*p21/norm(p21);
            pt1 = p2+p2t1;
            p2t2 = obj.radius/tan(0.5*theta)*p23/norm(p23);
            pt2 = p2+p2t2;
            
            pt1t2 = pt2-pt1;
            pt1m1 = 0.5*pt1t2;
            p2m1 = p2t1+pt1m1;
            p2c = obj.radius/sin(0.5*theta)/norm(p2m1)*p2m1;
            center = p2+p2c;            
        end
        
        %% Calculate Semicircle Transition
        function CalcSemicircleInfo(obj, pos)
            for idx=1:obj.numarc
                p1 = pos(:,2*idx-1); p2 = pos(:,2*idx); p3 = pos(:,2*idx+1); p4 = pos(:,2*idx+2);
                [obj.center(:,idx), pt1, pt2] = obj.CalcSemicirclePoints(p1,p2,p3,p4);
                obj.pt(:,2*idx-1) = pt1;
                obj.pt(:,2*idx) = pt2;
                obj.dr(idx) = pi*obj.radius;
                obj.rot{idx} = obj.CalcArcRot(obj.center(:,idx), pt1, pos(:,2*idx));
                if idx~=1
                    obj.dl(idx) = norm(obj.pt(:,2*idx-1)-obj.pt(:,2*idx-2));
                else
                    obj.dl(idx) = norm(pos(:,1)-obj.pt(:,1));
                end
                obj.line_vec(:,idx) = (pos(:,2*idx)-pos(:,2*idx-1))/norm(pos(:,2*idx)-pos(:,2*idx-1));
            end
            obj.dl(obj.numarc+1) = norm(pos(:,end)-obj.pt(:,end));
            obj.line_vec(:,obj.numarc+1) = (pos(:,end)-pos(:,end-1))/norm(pos(:,end)-pos(:,end-1));
            obj.distance = sum(obj.dl)+sum(obj.dr);
        end
        
        function [center, pt1, pt2] = CalcSemicirclePoints(obj, p1, p2, p3, p4)
            p21 = p1-p2; p34 = p4-p3;
            p2t1 = obj.radius*p21/norm(p21);
            p3t2 = obj.radius*p34/norm(p34);
            pt1 = p2+p2t1;
            pt2 = p3+p3t2;
            center = 0.5*(pt1+pt2);
        end
        
        %% Useful and Universal Functions
        function rot = CalcArcRot(obj, center, p1, p2)
            px = p1-center;
            n = px/norm(px);
            a = cross(px, (p2-center));
            a = a/norm(a);
            o = cross(a, n);
            rot = [n, o, a];
        end

        function dis = CalcDisInsterval(obj)
            dis = zeros(1,length(obj.dl)+length(obj.dr)+1);
            m = dis(1);
            for idx=2:length(dis)
                if mod(idx,2)==0
                    m = m+obj.dl(idx/2);
                else
                    m = m+obj.dr((idx-1)/2);
                end
                dis(idx) = m;
            end
        end
        
        function idx = CalcPosidx(obj, varp)
            if varp>obj.dis_interval(end)
                idx = length(obj.dis_interval)-1;
            elseif varp<obj.dis_interval(1)
                idx = 1;
            else
                idx = discretize(varp, obj.dis_interval);
            end
        end            
        
        %% Generate and Plot Trajectory 
        function p = GeneratePath(obj, varp)
            idx = obj.CalcPosidx(varp);
            if mod(idx,2)==1
                if idx==1
                    p = obj.p_initial+obj.line_vec(:,idx)*varp;
                else
                    p = obj.pt(:,idx-1)+obj.line_vec(:,(idx+1)/2)*(varp-obj.dis_interval(idx));
                end
            else
                th = (varp-obj.dis_interval(idx))/obj.radius;
                parc = zeros(3,1);
                parc(1) = obj.radius*cos(th);
                parc(2) = obj.radius*sin(th);
                p = obj.center(:,idx/2)+obj.rot{idx/2}*parc;
            end
        end
        
        function [p, v, a] = GenerateMotion(obj, varp, varv, vara)
            p = obj.GeneratePath(varp);
            idx = obj.CalcPosidx(varp);
            if mod(idx,2)==1
                v = varv*obj.line_vec(:,(idx+1)/2);
                a = vara*obj.line_vec(:,(idx+1)/2);
            else
                s = varp-obj.dis_interval(idx);
                r = obj.radius;
                vc=zeros(3,1); ac = zeros(3,1);
                vc(1) = -varv*sin(s/r);
                vc(2) = varv*cos(s/r);
                v = obj.rot{idx/2}*vc;
                ac(1) = -varv^2*cos(s/r)/r-vara*sin(s/r);
                ac(2) = -varv^2*sin(s/r)/r+vara*cos(s/r);
                a = obj.rot{idx/2}*ac;
            end
            v = [v; norm(v)]; a = [a;norm(a)];
        end
        
        function [pos, vel, acc] = GenerateTraj(obj, varp, varv, vara)
            pos = []; vel = []; acc = [];
            for idx=1:length(varp)
                [p, v, a] = obj.GenerateMotion(varp(idx), varv(idx), vara(idx));
                pos = [pos, p];
                vel = [vel, v];
                acc = [acc, a];
            end
        end
        
        function PlotTraj(obj, varp, varv, vara, tf, dt)
            [pos, vel, acc] = obj.GenerateTraj(varp, varv, vara);
            str_pos = {'px', 'py', 'pz'};
            str_vel = {'vx', 'vy', 'vz', '|v|'};
            str_acc = {'ax', 'ay', 'az', '|a|'};
            time=0:dt:tf;
            figure
            for idx=1:3
                subplot(3,1,idx); plot(time, pos(idx,:), 'k-'); grid on; ylabel(str_pos{idx});
            end
            figure
            for idx=1:4
                subplot(4,2,2*idx-1); plot(time, vel(idx,:), 'k-'); grid on; ylabel(str_vel{idx});
                subplot(4,2,2*idx); plot(time, acc(idx,:), 'k-'); grid on; ylabel(str_acc{idx});
            end    
        end
        
    end
      
end

File: src/cfiles/ArcTransPathPlanner.cpp
#include "ArcTransPathPlanner.h"

ArcTransPathPlanner::ArcTransPathPlanner(MatrixXd pos, double radius)
{
	InitPlanner(pos, radius);
}

void ArcTransPathPlanner::InitPlanner(MatrixXd pos, double radius)
{
	_p_initial = pos.col(0);
	_p_goal = pos.rightCols(1);
	_nump = pos.cols();
	if (radius>0)
	{
		_numarc = _nump-2;
		_radius = radius;
		CalcArcInfo(pos);
	}
	else
	{
		_numarc = _nump/2-1;
		_radius = 0.5*MathTools::Norm(pos.col(2)-pos.col(1));
		CalcSemicircleInfo(pos);
	}
	_dis_interval = CalcDisInterval();
}

void ArcTransPathPlanner::CalcArcInfo(MatrixXd pos)
{
	_center.setZero(3, _numarc);
	_theta.setZero(_numarc);
	_pt.setZero(3, 2*_numarc);
	_dr.setZero(_numarc);
	_dl.setZero(_numarc+1);
	_line_vec.setZero(3, _numarc+1);
	for (int idx = 0; idx<_numarc; idx++)
	{
		ArcData arc_data = CalcArcPoints(pos.col(idx), pos.col(idx+1), pos.col(idx+2));
		_center.col(idx) = arc_data.center;
		_theta(idx) = arc_data.theta;
		_pt.col(2*idx) = arc_data.pt1;
		_pt.col(2*idx+1) = arc_data.pt2;
		_dr(idx) = _radius*(pi-_theta(idx));
		_dl(idx) = MathTools::Norm(arc_data.pt1-pos.col(idx));
		Matrix3d tmp_rot= CalcArcRot(arc_data.center, arc_data.pt1, arc_data.pt2);
		_rot.push_back(tmp_rot);
		if (idx!=0)
			_dl(idx) = _dl(idx)-MathTools::Norm(pos.col(idx)-_pt.col(2*idx-1));
		if (idx==_numarc-1)
			_dl(idx+1) = MathTools::Norm(pos.rightCols(1)-arc_data.pt2);
		_line_vec.col(idx) = pos.col(idx+1)-pos.col(idx);
		_line_vec.col(idx).normalize();
	}
	_line_vec.col(_numarc) = pos.col(_nump-1)-pos.col(_nump-2);
	_line_vec.col(_numarc).normalize();
	_distance = _dl.sum()+_dr.sum();
}

Matrix3d ArcTransPathPlanner::CalcArcRot(Vector3d center, Vector3d p1, Vector3d p2)
{
	Matrix3d rot;
	Vector3d px = p1-center;
	Vector3d pz = px.cross(p2-center);
	Vector3d py = pz.cross(px);
	px.normalize(); py.normalize(); pz.normalize();
	rot<<px, py, pz;

	return rot;
}

ArcData ArcTransPathPlanner::CalcArcPoints(Vector3d p1, Vector3d p2, Vector3d p3)
{
	ArcData arc_data;
	Vector3d p21 = p1-p2, p23 = p3-p2;
	double tmp_cos = p21.dot(p23)/p21.norm()/p23.norm();
	MathTools::LimitNum(-1, tmp_cos, 1);
	double theta = acos(tmp_cos);

	Vector3d p2t1 = _radius/tan(0.5*theta)/p21.norm()*p21;
	Vector3d pt1 = p2+p2t1;
	Vector3d p2t2 = _radius/tan(0.5*theta)/p23.norm()*p23;
	Vector3d pt2 = p2+p2t2;

	Vector3d pt1t2 = pt2-pt1;
	Vector3d pt1m1 = 0.5*pt1t2;
	Vector3d p2m1 = p2t1+pt1m1;
	Vector3d p2c = _radius/sin(0.5*theta)/p2m1.norm()*p2m1;
	Vector3d center = p2+p2c;

	arc_data.center = center; arc_data.pt1 = pt1; arc_data.pt2 = pt2;
	arc_data.theta = theta;
	return arc_data;
}

void ArcTransPathPlanner::CalcSemicircleInfo(MatrixXd pos)
{
	_center.setZero(3, _numarc);
	_pt.setZero(3, 2*_numarc);
	_dr.setZero(_numarc);
	_dl.setZero(_numarc+1);
	_line_vec.setZero(3, _numarc+1);
	for (int idx = 0; idx<_numarc; idx++)
	{
		VectorXd p1, p2, p3, p4;
		p1 = pos.col(2*idx);
		p2 = pos.col(2*idx+1);
		p3 = pos.col(2*idx+2);
		p4 = pos.col(2*idx+3);
		ArcData arc_data = CalcSemicirclePoints(p1, p2, p3, p4);
		_center.col(idx) = arc_data.center;
		_pt.col(2*idx) = arc_data.pt1;
		_pt.col(2*idx+1) = arc_data.pt2;
		_dr(idx) = pi*_radius;
		Matrix3d tmp_rot = CalcArcRot(arc_data.center, arc_data.pt1, pos.col(2*idx+1));
		_rot.push_back(tmp_rot);
		if (idx!=0)
			_dl(idx) = MathTools::Norm(_pt.col(2*idx)-_pt.col(2*idx-1));
		else
			_dl(idx) = MathTools::Norm(pos.col(0)-_pt.col(0));
		_line_vec.col(idx) = (pos.col(2*idx+1)-pos.col(2*idx))/MathTools::Norm(pos.col(2*idx+1)-pos.col(2*idx));
	}
	_dl(_numarc) = MathTools::Norm(pos.rightCols(1)-_pt.rightCols(1));
	MatrixXd end_pos = pos.rightCols(2);
	_line_vec.col(_numarc) = (end_pos.col(1)-end_pos.col(0))/MathTools::Norm(end_pos.col(1)-end_pos.col(0));
	_distance = _dl.sum()+_dr.sum();
}

ArcData ArcTransPathPlanner::CalcSemicirclePoints(Vector3d p1, Vector3d p2, Vector3d p3, Vector3d p4)
{
	Vector3d p21, p34;
	p21 = p1-p2;
	p34 = p4-p3;
	Vector3d p2t1 = _radius*p21/p21.norm();
	Vector3d p3t2 = _radius*p34/p34.norm();
	Vector3d pt1 = p2+p2t1;
	Vector3d pt2 = p3+p3t2;
	Vector3d center = 0.5*(pt1+pt2);

	ArcData arc_data;
	arc_data.center = center;
	arc_data.pt1 = pt1;
	arc_data.pt2 = pt2;
	arc_data.theta = pi/2;

	return arc_data;
}

VectorXd ArcTransPathPlanner::CalcDisInterval()
{
	int n = _dl.size()+_dr.size()+1;
	VectorXd dis;
	dis.setZero(n);
	double m = dis(0);
	for (int idx = 1; idx<n; idx++)
	{
		if (idx%2==1)
			m += _dl((idx-1)/2);
		else
			m += _dr((idx-2)/2);
		dis(idx) = m;
	}
	return dis;
}

int ArcTransPathPlanner::CalcPosIdx(double varp)
{
	int idx = 0;
	int n = _dis_interval.size();
	if (varp>_dis_interval(n-1))
		idx = _dis_interval.size()-1;
	else if (varp<_dis_interval(0))
		idx = 0;
	else
		idx = MathTools::Discretize(_dis_interval, _dis_interval.size(), varp);

	return idx;
}

Vector3d ArcTransPathPlanner::GeneratePath(double varp)
{
	int idx = CalcPosIdx(varp);
	Vector3d p;
	if (idx%2==0)
	{
		if (idx==0)
			p = _p_initial+_line_vec.col(idx)*varp;
		else
			p = _pt.col(idx-1)+_line_vec.col(idx/2)*(varp-_dis_interval(idx));
	}
	else
	{
		double th = (varp-_dis_interval(idx))/_radius;
		Vector3d parc(0, 0, 0);
		parc(0) = _radius*cos(th);
		parc(1) = _radius*sin(th);
		p = _center.col((idx-1)/2)+_rot[(idx-1)/2]*parc;
	}

	return p;
}

RobotTools::CLineAVP ArcTransPathPlanner::GenerateMotion(double varp, double varv, double vara)
{
	RobotTools::CLineAVP avp;
	Vector3d p = GeneratePath(varp);
	Vector3d v(0, 0, 0), a(0, 0, 0);
	int idx = CalcPosIdx(varp);
	if (idx%2==0)
	{
		v = varv*_line_vec.col(idx/2);
		a = vara*_line_vec.col(idx/2);
	}
	else
	{
		double s = varp-_dis_interval(idx);
		double r = _radius;
		Vector3d vc(0, 0, 0), ac(0, 0, 0);
		vc(0) = -varv*sin(s/r);
		vc(1) = varv*cos(s/r);
		v = _rot[(idx-1)/2]*vc;
		ac(0) = -pow(varv, 2)*cos(s/r)/r-vara*sin(s/r);
		ac(1) = -pow(varv, 2)*sin(s/r)/r+vara*cos(s/r);
		a = _rot[(idx-1)/2]*ac;
	}

	avp.pos = p; avp.vel = v; avp.acc = a;
	return avp;
}

